crates/rune/src/macros/macro_compiler.rs

```diff
@@ -74,4 +76,67 @@ impl MacroCompiler<'_> {
 
         Ok(output)
     }
+
+    /// Compile the given macro into the given output type.
+    pub(crate) fn eval_attribute_macro<T>(
+        &mut self,
+        attribute: &ast::Attribute,
+        item: &ast::Item,
+    ) -> compile::Result<T>
+    where
+        T: Parse,
+    {
+        let span = attribute.span();
+
+        if !self.options.macros {
+            return Err(compile::Error::msg(
+                span,
+                "macros must be enabled with `-O macros=true`",
+            ));
+        }
+
+        // TODO: include information on the module the macro is being called
+        // from.
+        //
+        // TODO: Figure out how to avoid performing ad-hoc lowering here.
+        let arena = crate::hir::Arena::new();
+        let ctx = crate::hir::lowering::Ctx::new(&arena, self.query.borrow());
+        let path = crate::hir::lowering::path(&ctx, &attribute.path)?;
+        let named = self.query.convert_path(self.context, &path)?;
+
+        let hash = self.query.pool.item_type_hash(named.item);
+
+        let handler = match self.context.lookup_attribute_macro(hash) {
+            Some(handler) => handler,
+            None => {
+                return Err(compile::Error::new(
+                    span,
+                    CompileErrorKind::MissingMacro {
+                        item: self.query.pool.item(named.item).to_owned(),
+                    },
+                ));
+            }
+        };
+
+        // TODO ATTRMCRO this should maybe remove the leading `=` or surrounding delimiter
+        let input_stream = &attribute.input;
```

Comment:
Hm, yeah. This would require some light parsing which is annoying. In Rust the AST is made up of [groups](https://doc.rust-lang.org/stable/proc_macro/enum.TokenTree.html#variant.Group) which are easy enough to strip.

Did you ever try what happens with attribute macros using `=` in Rust? I haven't checked that out yet.

Reply:
So it seems like key-value attributes are not supported in Rust:

![image](https://github.com/rune-rs/rune/assets/111092/82b87dfc-b8eb-4ee2-ba8d-502d54dc59a5)

Reply:
I think this might be an instance where it's worth deviating from Rust. The restrictions imposed by key-value attribute macros seems to have been [a historical accident](https://github.com/rust-lang/rust/issues/34981). Us having to reparse macro input just to clean it up seems silly.

If so, it seems natural to parse on the input as you're already providing it since it leads to unambiguous parsing requirements across styles. To facilitate parsing it would be a good idea to implement a helper function for parsing groups, [which implements this balancing logic](https://github.com/rune-rs/rune/blob/main/crates/rune/src/ast/macro_call.rs#L74).

What do you think?

Reply:
makes sense

Reply:
I could also imagine having an AST element for a delimited group that one could use to just call parse with.